### DISCUSSION TITLE
Unexpected response

### DISCUSSION THREAD

**Discussion by felrock** [miguelgrinberg/python-socketio#858] [Q&A] [answered]
**Describe the bug**
Running the client/server examples inside a docker gives me an 'Unexpected response from server' error. Using Python 3.8.

**To Reproduce**
Run examples as simple_server.py and simple_client.py on the same machine, except that simple_client is in a docker env.

https://python-socketio.readthedocs.io/en/latest/intro.html#client-examples

https://python-socketio.readthedocs.io/en/latest/intro.html#server-examples

**Expected behavior**
Terminal print that a connection was established.

**Logs**
Client:
Traceback (most recent call last):
  File "simple_client.py", line 18, in <module>
    sio.connect('http://127.0.0.1:5000')
  File "/usr/local/lib/python3.8/dist-packages/socketio/client.py", line 338, in connect
    raise exceptions.ConnectionError(exc.args[0]) from None
socketio.exceptions.ConnectionError: Unexpected response from server

Server:
(2176514) wsgi starting up on http://127.0.0.1:5000
(2176514) accepted ('127.0.0.1', 42518)
connect  8b71ba2ba70f4ec3bd0491d76fcaba22
127.0.0.1 - - [25/Jan/2022 13:15:15] "GET /socket.io/?transport=polling&EIO=4&t=1643112915.2445767 HTTP/1.1" 200 349 0.000694

**Comment by miguelgrinberg** [comment]
Please don't file issues against this project when you have a question about why your application does not work. The issues board is to track actual issues, while the discussion board is to troubleshoot or ask for help.

If you are running the client inside a container and the server on the real host, then the address 127.0.0.1 issued in the container refers to the container itself, not the Docker host. The hostname `host.docker.internal` is in general mapped to the Docker host. Try to connect to `http://host.docker.internal:5000`.

**Reply by felrock** [reply]
Got it!

**Comment by felrock** [comment]
Then i get the following error

```
>>> client.connect('http://host.docker.internal:5000')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/dist-packages/socketio/client.py", line 338, in connect
    raise exceptions.ConnectionError(exc.args[0]) from None
socketio.exceptions.ConnectionError: HTTPConnectionPool(host='host.docker.internal', port=5000): Max retries exceeded with url: /socket.io/?transport=polling&EIO=4&t=1643134495.1887655 (Caus
ed by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f159c4409a0>: Failed to establish a new connection: [Errno -2] Name or service not known'))
```

 A simple wget shows me that i can access the host localhost

```
root@devfelix:/src/workspace# wget 127.0.0.1:5000
--2022-01-25 18:15:14--  http://127.0.0.1:5000/
Connecting to 127.0.0.1:5000... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1135 (1.1K) [text/html]
Saving to: 'index.html'

index.html                                      100%[=====================================================================================================>]   1.11K  --.-KB/s    in 0s      

2022-01-25 18:15:14 (228 MB/s) - 'index.html' saved [1135/1135]
```

**Reply by miguelgrinberg** [reply]
I'm sorry but I have no context to understand what you are showing me. I have no idea what container(s) you have running, which process do you have in each, and from where you are connecting to where.

**Reply by felrock** [reply]
Yeah that's my bad, I made a simpler scenario test. Now there is no docker container, only a single conda environment. Running python 3.7.11 and it has python-socketio installed. Both client and server the first examples on the readthedocs site.

server.py: https://python-socketio.readthedocs.io/en/latest/intro.html#server-examples
client.py: https://python-socketio.readthedocs.io/en/latest/intro.html#client-examples

output from running server.py:
(conda_env) user@comp:~/example$ python server.py 
(2215153) wsgi starting up on http://0.0.0.0:5000

output from running client.py:
(conda_env) user@comp:~/example$ python client.py 
Traceback (most recent call last):
  File "client.py", line 18, in <module>
    sio.connect('http://localhost:5000')
  File "/home/felix/anaconda3/envs/socketio/lib/python3.7/site-packages/socketio/client.py", line 338, in connect
    raise exceptions.ConnectionError(exc.args[0]) from None
socketio.exceptions.ConnectionError: Unexpected response from server

**Reply by felrock** [reply]
I get the same if only keep the connect/disconnect events in both scripts

**Reply by miguelgrinberg** [reply]
Any chance your port 5000 is already used by some other thing on your system? Replace 5000 with 5001 on both scripts and try again to rule that out.

**Reply by felrock** [reply]
That was it! Thanks for you help.

**Reply by felrock** [reply]
Sorry for flip-flopping like this, but the issue returned so i traced it down to engineio and to r.content. I removed the try-except clause and it turns out my client does recieve a 200 ack, but python has trouble in decoding it to utf-8.

Error log:

```
   Traceback (most recent call last):
  File "/src/workspace/build/web_connection/web_connection/socket_handler.py", line 51, in __init__
    self.sio.connect('http://127.0.0.1:5000')
  File "/usr/local/lib/python3.8/dist-packages/socketio/client.py", line 331, in connect
    self.eio.connect(real_url, headers=real_headers,
  File "/usr/local/lib/python3.8/dist-packages/engineio/client.py", line 185, in connect
    return getattr(self, '_connect_' + self.transports[0])(
  File "/usr/local/lib/python3.8/dist-packages/engineio/client.py", line 298, in _connect_polling
    p = payload.Payload(encoded_payload=r.content.decode('utf-8'))
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 4: invalid start byte

```
r.content printed:

`b'\x00\x01\x00\t\xff0{"sid":"032b6ff2bd164cc2bba216278bdb4557","upgrades":["websocket"],"pingTimeout":60000,"pingInterval":25000}\x00\x02\xff40'`

It looks the same on a request that "works" and one that creates the Unexpected response from server error that i saw initially.

This version of engineio seems to break when it tries to decode, _python_engineio-4.3.1-py3-none-any.whl_

**Reply by felrock** [reply]
Maybe this should be a discussion on engineio repo?

**Comment by felrock** [comment, marked as answer]
After getting the issue again it turned out that the issue was a old version of python-engineio, updating to later version fixed it.
